File: reference/permissions.py
"""Role, object-level and field-level access control (spec sections 5, 13, I, N)."""
from dataclasses import dataclass, field
import events


class AccessDenied(Exception):
    pass


@dataclass
class Actor:
    user_id: str
    role: str
    name: str
    office: str = "RW"
    scope: dict = field(default_factory=dict)   # e.g. {"customer_id": ..., "shipment_ids": [...]}


# (role -> set of "resource:action") -------------------------------------------------
ROLE_GRANTS = {
    "ceo": {"*:*"},
    "venture_manager": {
        "lead:*", "request:*", "project:*", "task:*", "quotation:*", "order:*",
        "customer:read", "supplier:read", "shipment:read", "shipment:create",
        "package:read", "package:update", "invoice:read", "payment:read",
        "sourcing_task:create",
    },
    "china_sourcing": {
        "supplier:*", "rfq:*", "supplier_quote:*", "po:*", "visit:create",
        "inspection:read", "capa:*", "project:read", "task:read", "package:read",
    },
    "china_warehouse": {
        "visit:read", "visit:accept", "inspection:*", "package:*", "receipt:*",
        "supplier:read", "po:read", "capa:read",
    },
    "front_office": {
        "customer:read", "call:*", "promise:*", "quotation:read", "quotation:draft",
        "project:read", "order:read", "shipment:read", "petty_cash:*",
    },
    "finance": {
        "invoice:*", "payment:*", "po:approve", "margin:read", "commission:*",
        "project:read", "order:read", "supplier:read",
    },
    "sales_agent": {
        "lead:create", "lead:read", "customer:create", "customer:read",
        "request:create", "quotation:read", "order:read", "commission:read",
    },
    "customer": {
        "project:read", "quotation:read", "order:read", "invoice:read",
        "payment:create", "shipment:read", "package:read",
    },
    "logistics_partner": {
        "shipment:read", "package:read", "delivery:*", "customs_doc:*",
    },
}

# Fields nobody outside this set may ever see (spec section I: "Never expose full
# margin calculations to unauthorized users"; section N: Imari sees no costs).
CONFIDENTIAL_FIELDS = {
    "supplier_unit_cost": {"ceo", "china_sourcing", "venture_manager", "finance"},
    "po_total":           {"ceo", "china_sourcing", "venture_manager", "finance"},
    "target_price":       {"ceo", "china_sourcing", "venture_manager", "finance"},
    "walkaway_price":     {"ceo", "china_sourcing", "venture_manager", "finance"},
    "margin_pct":         {"ceo", "venture_manager", "finance"},
    "agent_commission":   {"ceo", "finance", "sales_agent"},
}


def can(actor, resource, action):
    grants = ROLE_GRANTS.get(actor.role, set())
    return any(g in grants for g in ("*:*", f"{resource}:*", f"*:{action}", f"{resource}:{action}"))


def _in_scope(actor, obj):
    """Object-level access: least privilege (spec section 5)."""
    if actor.role in ("ceo", "venture_manager", "finance", "china_sourcing",
                      "china_warehouse", "front_office"):
        return True
    if actor.role == "customer":
        return getattr(obj, "customer_id", None) == actor.scope.get("customer_id")
    if actor.role == "sales_agent":
        return getattr(obj, "agent_id", None) == actor.user_id or \
               getattr(obj, "customer_id", None) in actor.scope.get("customer_ids", [])
    if actor.role == "logistics_partner":
        # A partner is scoped to shipments. The object may BE a shipment, or hang
        # off one (package, delivery, customs doc), so resolve to a shipment ref.
        sid = (getattr(obj, "shipment_ref", None)
               or getattr(obj, "shipment_id", None)
               or (obj.ref if obj.__class__.__name__ == "Shipment" else None))
        return sid in actor.scope.get("shipment_ids", [])
    return False


def authorize(actor, resource, action, obj=None):
    if not can(actor, resource, action):
        events.audit(actor, "ACCESS_DENIED", f"{resource}:{action}", "role grant missing")
        raise AccessDenied(f"{actor.role} may not {action} {resource}")
    if obj is not None and not _in_scope(actor, obj):
        events.audit(actor, "ACCESS_DENIED", f"{resource}:{action}", "object out of scope")
        raise AccessDenied(f"{actor.role} {actor.user_id} is out of scope for that {resource}")
    events.audit(actor, f"{resource}:{action}", getattr(obj, "ref", "-"))
    return True


def project_fields(actor, record: dict) -> dict:
    """Field-level masking applied on read."""
    out = {}
    for k, v in record.items():
        allowed = CONFIDENTIAL_FIELDS.get(k)
        if allowed is None or actor.role in allowed:
            out[k] = v
        else:
            out[k] = "***"
    return out

File: packages/contracts/src/permissions.ts
/**
 * Role grants, object scope and field masking.
 * Authorisation happens at the SERVICE layer, because object scope depends on
 * the record being touched, not on the route.
 *
 * `customer` is deliberately NOT a login role here. UZA Nexus OS is the internal operating
 * layer — a customer never authenticates into it. Customer-facing access (their own
 * quotation, order status, payment) belongs on uzabulk.com, reading Nexus as its source of
 * truth through a service integration, not by handing a customer a Nexus login. The
 * `Customer` business record (packages/contracts and the trade chain) is unaffected —
 * this removes only the login role, not the record every order/quotation still points at.
 */
export type Role =
  | 'ceo'
  | 'venture_manager'
  | 'china_sourcing'
  | 'china_warehouse'
  | 'front_office'
  | 'finance'
  | 'sales_agent'
  | 'logistics_partner';

export interface Actor {
  readonly userId: string;
  readonly role: Role;
  readonly office: string;
  readonly scope: {
    customerId?: string;
    customerIds?: readonly string[];
    shipmentRefs?: readonly string[];
  };
}

export const ROLE_GRANTS: Record<Role, readonly string[]> = {
  ceo: ['*:*'],
  // `uza-id:resolve`/`link`/`links` on the six internal-staff roles below, and the
  // additional `uza-id:merge`/`consent` on venture_manager, are a policy call made
  // when the module went from zero authorization to enforced — there was no prior
  // grant to recover, so this is a proposal, not a fact. Two tiers, deliberately:
  // resolve/link/links is the routine cross-system integration path every satellite
  // system calls to onboard or look up a person — open to internal staff, never to
  // `customer` or `logistics_partner` (the two external, expiry-limited account
  // kinds). merge/consent are irreversible-ish (a merge folds two people together;
  // consent controls a Law N°058/2021 flag) and stay limited to the two roles this
  // codebase already treats as trusted-admin (`ceo` via `*:*`; `venture_manager`,
  // the other role with broad `*` grants below). Confirm this with the founder.
  venture_manager: [
    'lead:*',
    'request:*',
    'project:*',
    'task:*',
    'quotation:*',
    'order:*',
    'customer:read',
    'supplier:read',
    'shipment:read',
    'shipment:create',
    'package:read',
    'package:update',
    'invoice:read',
    'payment:read',
    'uza-id:resolve',
    'uza-id:link',
    'uza-id:links',
    'uza-id:merge',
    'uza-id:consent',
  ],
  china_sourcing: [
    'supplier:*',
    'rfq:*',
    'supplierQuote:*',
    'po:*',
    'visit:create',
    'inspection:read',
    'capa:*',
    'project:read',
    'task:read',
    'package:read',
    'uza-id:resolve',
    'uza-id:link',
    'uza-id:links',
  ],
  china_warehouse: [
    'visit:read',
    'visit:accept',
    'inspection:*',
    'package:*',
    'receipt:*',
    'supplier:read',
    'po:read',
    'capa:read',
    'uza-id:resolve',
    'uza-id:link',
    'uza-id:links',
  ],
  front_office: [
    'customer:read',
    'call:*',
    'promise:*',
    'quotation:read',
    'quotation:draft',
    'project:read',
    'order:read',
    'shipment:read',
    'pettyCash:*',
    // Added when the `customer` login role was removed (30 Aug 2026): a customer never
    // authenticates into Nexus, so someone on staff has to be the one who records that a
    // payment came in — front_office is the customer-facing coordination role (already
    // holds customer:read, call:*, promise:*), receiving proof by call/WhatsApp/email and
    // logging it here. Finance still holds sole verification authority (payment:*) — this
    // only lets front_office create the unverified record; rule 1 (payment gates
    // procurement, verified by Finance) is untouched.
    'payment:create',
    'uza-id:resolve',
    'uza-id:link',
    'uza-id:links',
  ],
  finance: [
    'invoice:*',
    'payment:*',
    'installment:*',
    'po:approve',
    'margin:read',
    'commission:*',
    'project:read',
    'order:read',
    'supplier:read',
    'claim:*',
    'uza-id:resolve',
    'uza-id:link',
    'uza-id:links',
  ],
  sales_agent: [
    'lead:create',
    'lead:read',
    'customer:create',
    'customer:read',
    'request:create',
    'quotation:read',
    'order:read',
    'commission:read',
    'uza-id:resolve',
    'uza-id:link',
    'uza-id:links',
  ],
  logistics_partner: ['shipment:read', 'package:read', 'delivery:*', 'customsDoc:*'],
};

/**
 * Masked on READ, not filtered in the UI. A field absent from this map is visible
 * to anyone whose role grant already allows the resource.
 */
export const CONFIDENTIAL_FIELDS: Record<string, readonly Role[]> = {
  supplierUnitCost: ['ceo', 'china_sourcing', 'venture_manager', 'finance'],
  poTotal: ['ceo', 'china_sourcing', 'venture_manager', 'finance'],
  targetPrice: ['ceo', 'china_sourcing', 'venture_manager', 'finance'],
  walkawayPrice: ['ceo', 'china_sourcing', 'venture_manager', 'finance'],
  marginPct: ['ceo', 'venture_manager', 'finance'],
  realizedMargin: ['ceo', 'venture_manager', 'finance'],
  agentCommission: ['ceo', 'finance', 'sales_agent'],
  // Freight monetary figures on a Shipment — a logistics_partner sees weight/CBM but
  // never cost (constitution rule 12). measuredRevenueTon is a volumetric the China
  // warehouse legitimately needs; only the money figures are strictly cost.
  freightPaidMinor: ['ceo', 'venture_manager', 'finance'],
  billedRevenueTon: ['ceo', 'venture_manager', 'finance'],
  measuredRevenueTon: ['ceo', 'venture_manager', 'finance', 'china_warehouse'],
};

export const MASK = '***' as const;

export const can = (actor: Actor, resource: string, action: string): boolean => {
  const grants = ROLE_GRANTS[actor.role];
  return ['*:*', `${resource}:*`, `*:${action}`, `${resource}:${action}`].some((g) =>
    grants.includes(g),
  );
};

/**
 * Object scope. Resolve explicitly per role — never guess from attribute names.
 * Guessing was a real defect in the reference implementation: a partner scoped to
 * shipments reaches a delivery, a package and a customs doc by different paths.
 */
export interface Scopable {
  customerId?: string;
  agentId?: string;
  shipmentRef?: string;
  ref?: string;
  kind?: string;
}

export const inScope = (actor: Actor, obj: Scopable): boolean => {
  switch (actor.role) {
    case 'ceo':
    case 'venture_manager':
    case 'finance':
    case 'china_sourcing':
    case 'china_warehouse':
    case 'front_office':
      return true;
    case 'sales_agent':
      return (
        obj.agentId === actor.userId ||
        (!!obj.customerId && (actor.scope.customerIds ?? []).includes(obj.customerId))
      );
    case 'logistics_partner': {
      const ref = obj.shipmentRef ?? (obj.kind === 'shipment' ? obj.ref : undefined);
      return !!ref && (actor.scope.shipmentRefs ?? []).includes(ref);
    }
  }
};

export const maskFields = <T extends Record<string, unknown>>(
  actor: Actor,
  record: T,
): { [K in keyof T]: T[K] | typeof MASK } => {
  const out = {} as { [K in keyof T]: T[K] | typeof MASK };
  for (const key of Object.keys(record) as (keyof T)[]) {
    const allowed = CONFIDENTIAL_FIELDS[key as string];
    out[key] = !allowed || allowed.includes(actor.role) ? record[key] : MASK;
  }
  return out;
};
